src/x86/runtime.rs

```diff
@@ -0,0 +1,121 @@
+#[repr(u8)]
+pub enum __Feature {
+    /// SSE (Streaming SIMD Extensions)
+    sse = 0,
+    /// SSE2 (Streaming SIMD Extensions 2)
+    sse2 = 1,
+    /// SSE3 (Streaming SIMD Extensions 3)
+    sse3 = 2,
+    /// SSSE3 (Supplemental Streaming SIMD Extensions 3)
+    ssse3 = 3,
+    /// SSE4.1 (Streaming SIMD Extensions 4.1)
+    sse4_1 = 4,
+    /// SSE4.2 (Streaming SIMD Extensions 4.2)
+    sse4_2 = 5,
+    /// AVX (Advanced Vector Extensions)
+    avx = 6,
+    /// AVX2 (Advanced Vector Extensions 2)
+    avx2 = 7,
+    /// FMA (Fused Multiply Add)
+    fma = 8,
+    /// BMI1 (Bit Manipulation Instructions 1)
+    bmi1 = 9,
+    /// BMI1 (Bit Manipulation Instructions 2)
+    bmi2 = 10,
+    /// ABM (Advanced Bit Manipulation) on AMD / LZCNT (Leading Zero Count) on Intel
+    abm = 11,
+    /// TBM (Trailing Bit Manipulation)
+    tbm = 12,
+    /// POPCNT (Population Count)
+    popcnt = 13,
+
+    #[doc(hidden)]
+    __NonExhaustive
+}
+
+fn set_bit(x: usize, bit: u32) -> usize {
+    debug_assert!(32 > bit);
+    x | 1 << bit
+}
+
+fn test_bit(x: usize, bit: u32) -> bool {
+    debug_assert!(32 > bit);
+    x & (1 << bit) != 0
+}
+
+fn inv_test_bit(v: usize, idx: u32) -> bool {
+    debug_assert!(32 > idx);
+    ((v >> idx) & 1) != 0
+}
+
+fn detect_features() -> usize {
+    let ebx;
+    let ecx;
+    let edx;
+
+    unsafe {
+        asm!("cpuid" : "={ecx}"(ecx), "={edx}"(edx)
+             : "{eax}"(0x00000001u32), "{ecx}"(0 as u32) : :);
+        asm!("cpuid" : "={ebx}"(ebx) : "{eax}"(0x00000007u32), "{ecx}"(0 as u32) : :);
+    }
+
+    let mut value: usize = 0;
+
+    // CPUID call with EAX=7, ECX=0 => Extended Features in EBX and ECX (unneeded):
+    if inv_test_bit(ebx, 3) { value = set_bit(value, __Feature::bmi1 as u32); }
+    if inv_test_bit(ebx, 5) { value = set_bit(value, __Feature::avx2 as u32); }
+    if inv_test_bit(ebx, 8) { value = set_bit(value, __Feature::bmi2 as u32); }
+
+    // CPUID call with EAX=1 => feature bits in ECX and EDX:
+    if inv_test_bit(ecx, 0) { value = set_bit(value, __Feature::sse3 as u32); }
+    if inv_test_bit(ecx, 5) { value = set_bit(value, __Feature::abm as u32); }
+    if inv_test_bit(ecx, 9) { value = set_bit(value, __Feature::ssse3 as u32); }
+    if inv_test_bit(ecx, 12) { value = set_bit(value, __Feature::fma as u32); }
+    if inv_test_bit(ecx, 19) { value = set_bit(value, __Feature::sse4_1 as u32); }
+    if inv_test_bit(ecx, 20) { value = set_bit(value, __Feature::sse4_2 as u32); }
+    if inv_test_bit(ecx, 21) { value = set_bit(value, __Feature::tbm as u32); }
+    if inv_test_bit(ecx, 23) { value = set_bit(value, __Feature::popcnt as u32); }
+    if inv_test_bit(ecx, 28) { value = set_bit(value, __Feature::avx as u32); }
+
+    if inv_test_bit(edx, 25) { value = set_bit(value, __Feature::sse as u32); }
+    if inv_test_bit(edx, 26) { value = set_bit(value, __Feature::sse2 as u32); }
+
+    value
+}
+
+static mut FEATURES: AtomicUsize = AtomicUsize::new(::std::usize::MAX);
```

Comment:
This can just be a plain `static` instead of `static mut`, removing the need for `unsafe` below

Reply:
Done.